src/dragonball/src/device_manager/vhost_user_net_dev_mgr.rs

```diff
@@ -0,0 +1,325 @@
+    DuplicatedUdsPath(String),
+    /// Failure from Virtio subsystem.
+    #[error(transparent)]
+    Virtio(VirtioError),
+    /// The update is not allowed after booting the microvm.
+    #[error("update operation is not allowed after boot")]
+    UpdateNotAllowedPostBoot,
+    /// Split this at some point.
+    /// Internal errors are due to resource exhaustion.
+    /// Users errors are due to invalid permissions.
+    #[error("cannot create a vhost-user-net device: {0:?}")]
+    CreateNetDevice(VirtioError),
+    /// Cannot initialize a MMIO Network Device or add a device to the MMIO Bus.
+    #[error("failure while registering network device: {0:?}")]
+    RegisterNetDevice(DeviceMgrError),
+}
+/// Configuration information for vhost user net devices.
+#[derive(Clone, Debug, Default, Deserialize, PartialEq, Serialize)]
+pub struct VhostUserNetDeviceConfigInfo {
+    /// vhost-user socket path.
+    pub sock_path: String,
+    /// Number of virtqueues to use.
+    pub num_queues: usize,
+    /// Size of each virtqueue.
+    pub queue_size: u16,
+    /// mac of the interface.
+    pub guest_mac: Option<MacAddr>,
+    /// Use shared irq
+    pub use_shared_irq: Option<bool>,
+    /// Use generic irq
+    pub use_generic_irq: Option<bool>,
+}
+
+impl VhostUserNetDeviceConfigInfo {
+    /// Returns a reference to the mac address. If the mac address is not configured, it
+    /// returns None.
+    pub fn guest_mac(&self) -> Option<&MacAddr> {
+        self.guest_mac.as_ref()
+    }
+
+    ///Rx and Tx queue and max queue sizes
+    pub fn queue_sizes(&self) -> Vec<u16> {
+        let mut queue_size = self.queue_size;
+        if queue_size == 0 {
+            queue_size = DEFAULT_QUEUE_SIZE;
+        }
+        let num_queues = if self.num_queues > 0 {
+            self.num_queues
+        } else {
+            DEFAULT_NUM_QUEUES
+        };
+        (0..num_queues).map(|_| queue_size).collect::<Vec<u16>>()
+    }
+}
+
+impl ConfigItem for VhostUserNetDeviceConfigInfo {
+    type Err = VhostUserNetDeviceError;
+
+    fn check_conflicts(&self, _other: &Self) -> std::result::Result<(), Self::Err> {
+        Ok(())
+    }
+
+    fn id(&self) -> &str {
+        &self.sock_path
+    }
+}
+
+/// Device manager to manage all vhost user net devices.
+pub struct VhostUserNetDeviceMgr {
+    pub(crate) configs: DeviceConfigInfos<VhostUserNetDeviceConfigInfo>,
+}
+
+impl VhostUserNetDeviceMgr {
+    fn create_device(
+        cfg: &VhostUserNetDeviceConfigInfo,
+        ctx: &mut DeviceOpContext,
+    ) -> Result<Box<VhostUserNet<GuestAddressSpaceImpl>>, VirtioError> {
+        match ctx.epoll_mgr.as_ref() {
+            Some(epoll_mgr) => Ok(Box::new(VhostUserNet::new_server(
+                &cfg.sock_path,
+                cfg.guest_mac(),
+                Arc::new(cfg.queue_sizes()),
+                epoll_mgr.clone(),
+            )?)),
+            None => Err(VirtioError::InvalidInput),
```

Comment:
how about making this error information more explicit like `NoEpollMgr`?

Reply:
I think it's great. However, other device managers, e.g. `VirtioNetDeviceMgr`, are still using the `InvalidInput` as well. Shall we file an issue to track this and fix this in another PR?